### PR TITLE
Implementation of Path::self()

### LINKED ISSUE #2282: Add Path::self()
Add an easy way to obtain the executable path. win/linux/mac only for now.
https://stackoverflow.com/a/1024937/205386

### PR DESCRIPTION
**Pull request by matejk**
This PR contains content of PR #3758 by @JackyWoo with minor additional changes.

Resolves #2282.

### CODE REVIEW

File: Foundation/src/Path_UNIX.cpp [hunk 1]
```diff
@@ -19,19 +19,64 @@
...
+#elif POCO_OS == POCO_OS_SOLARIS
+	char * execName = getexecname();
+	if (execName)
+		path = execName;
+#elif POCO_OS == POCO_OS_LINUX || POCO_OS == POCO_OS_ANDROID
+	const std::size_t size = sizeof(buf);
+	int n = readlink("/proc/self/exe", buf, size);
+	if (n > 0 && n < PATH_MAX)
+		path = buf;
+#endif
```
Comment (aleks-f) [inline Foundation/src/Path_UNIX.cpp:75]: let's throw `Poco::NotImplementedException` here

File: Foundation/testsuite/src/PathTest.cpp [hunk 2]
```diff
@@ -1619,6 +1619,27 @@ void PathTest::testWindowsSystem()
...
+	|| POCO_OS == POCO_OS_LINUX                               \
+	|| POCO_OS == POCO_OS_ANDROID
+
+	assertTrue(!self.empty());
+	Path p(self);
+
+	assertTrue(p.isAbsolute());
+	assertTrue(p.isFile());
+	assertTrue(self.find("testrunner") != std::string::npos);
+
```
Comment (aleks-f) [inline Foundation/testsuite/src/PathTest.cpp:1639]:
```c+=
#else
   std::cout << "Not implemented";
```

**Review by aleks-f — CHANGES_REQUESTED** [review]

**Comment by matejk** [comment]
@aleks-f , it is very unlikely that the functions to get current program name would fail, however: shall an exception be thrown in that case?

File: Foundation/testsuite/src/PathTest.cpp (hunk 2 above)
Comment (matejk) [reply Foundation/testsuite/src/PathTest.cpp]: Added tests for Windows and message for platforms where functionality is not implemented.

File: Foundation/src/Path_UNIX.cpp (hunk 1 above)
Comment (matejk) [reply Foundation/src/Path_UNIX.cpp:77]: Added throw.

File: Foundation/testsuite/src/PathTest.cpp [hunk 3]
```diff
@@ -1624,19 +1624,22 @@ void PathTest::testSelf()
...
 
 	assertTrue(!self.empty());
 	Path p(self);
 
 	assertTrue(p.isAbsolute());
 	assertTrue(p.isFile());
 	assertTrue(self.find("testrunner") != std::string::npos);
-
+#else
+	std::cout "Path::self() not implemented for this platform."
```
Comment (aleks-f) [inline Foundation/testsuite/src/PathTest.cpp:1642]: missing `<<`

**Review by aleks-f — CHANGES_REQUESTED** [review]

**Comment by aleks-f** [comment]
> @aleks-f , it is very unlikely that the functions to get current program name would fail, however: shall an exception be thrown in that case?

yes, throw something apropriate

**Review by aleks-f — APPROVED** [review]

**Pull request merged**